subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/execution/Interpreter.kt

```diff
@@ -0,0 +1,559 @@
+
+import java.io.File
+
+import java.lang.reflect.InvocationTargetException
+
+
+/**
+ * An optimised interpreter for the Kotlin DSL based on the idea of
+ * [partial evaluation](https://en.wikipedia.org/wiki/Partial_evaluation).
+ *
+ * Instead of interpreting a given Kotlin DSL script directly, the interpreter emits a
+ * specialized program that captures the optimal execution procedure for the particular
+ * combination of script structure (does it contain a `buildscript` block? a `plugins` block?
+ * a script body? etc), target object and context (top-level or not).
+ *
+ * The specialized program is then cached via a cheap cache key based on the original,
+ * unprocessed contents of the script, the target object type and parent `ClassLoader`.
+ *
+ * Because each program is specialized to a given script structure, a lot of work is
+ * avoided. For example, a top-level script containing a `plugins` block but no body
+ * can be compiled down to a specialized program that instantiates the precompiled
+ * `plugins` block class directly - without reflection - and does nothing else. The
+ * same strategy can be used for a **script plugin** (a non top-level script) with a
+ * body but no `buildscript` block since the classpath is completely determined at
+ * the time the specialized program is emitted.
+ *
+ * @see Program
+ * @see PartialEvaluator
+ * @see ResidualProgram
+ * @see ResidualProgramCompiler
+ */
+class Interpreter(val host: Host) {
+
+    interface Host {
+
+        fun cachedClassFor(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader
+        ): Class<*>?
+
+        fun cache(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader,
+            specializedProgram: Class<*>
+        )
+
+        fun cachedDirFor(
+            templateId: String,
+            sourceHash: HashCode,
+            parentClassLoader: ClassLoader,
+            initializer: (File) -> Unit
+        ): File
+
+        fun compilationClassPathOf(
+            classLoaderScope: ClassLoaderScope
+        ): ClassPath
+
+        fun loadClassInChildScopeOf(
+            classLoaderScope: ClassLoaderScope,
+            childScopeId: String,
+            location: File,
+            className: String,
+            accessorsClassPath: ClassPath?
+        ): Class<*>
+
+        fun applyPluginsTo(scriptHost: KotlinScriptHost<*>, pluginRequests: PluginRequests)
+
+        fun applyBasePluginsTo(project: Project)
+
+        fun closeTargetScopeOf(scriptHost: KotlinScriptHost<*>)
+
+        val implicitImports: List<String>
+    }
+
+    fun eval(
+        target: Any,
+        scriptSource: ScriptSource,
+        sourceHash: HashCode,
+        scriptHandler: ScriptHandler,
+        targetScope: ClassLoaderScope,
+        baseScope: ClassLoaderScope,
+        topLevelScript: Boolean,
+        options: EvalOptions = defaultEvalOptions
+    ) {
```

Comment:
When you've reached 8 arguments it might be time to consider refactoring your args into a class/builder.
Something like `EvalArgs`.
I notice that the `evaluate` above has pretty much the same arguments as the `eval` here. If you end up needing more arguments, this is just going to mean two places to update.